kcoin/genesis/genesis_generator_test.go

```diff
@@ -0,0 +1,300 @@
+			InvalidCommandFromValid: func(command GenesisOptions) GenesisOptions {
+				command.WalletAddressGenesisValidator = "0xe2ac86cbae1bbbb47d157516d334e70859a1be"
+				return command
+			},
+			ExpectedError: ErrInvalidWalletAddressValidator,
+		},
+		{
+			TestName: "Invalid wallet address less than 20 bytes without Hex prefix",
+			InvalidCommandFromValid: func(command GenesisOptions) GenesisOptions {
+				command.WalletAddressGenesisValidator = "e2ac86cbae1bbbb47d157516d334e70859a1be"
+				return command
+			},
+			ExpectedError: ErrInvalidWalletAddressValidator,
+		},
+		{
+			TestName: "Empty prefunded accounts",
+			InvalidCommandFromValid: func(command GenesisOptions) GenesisOptions {
+				command.PrefundedAccounts = []PrefundedAccount{}
+				return command
+			},
+			ExpectedError: ErrEmptyPrefundedAccounts,
+		},
+		{
+			TestName: "Prefunded accounts does not include validator address",
+			InvalidCommandFromValid: func(command GenesisOptions) GenesisOptions {
+				command.PrefundedAccounts = []PrefundedAccount{
+					{
+						WalletAddress: "0xaaaaaacbae1bbbb47d157516d334e70859a1bee4",
+						Balance:       "15",
+					},
+				}
+				return command
+			},
+			ExpectedError: ErrWalletAddressValidatorNotInPrefundedAccounts,
+		},
+		{
+			TestName: "Prefunded accounts has invalid account.",
+			InvalidCommandFromValid: func(command GenesisOptions) GenesisOptions {
+				command.PrefundedAccounts = []PrefundedAccount{
+					{
+						WalletAddress: "0xe2ac86cbae1bbbb47d157516d334e70859a1bee4",
+						Balance:       "15",
+					},
+					{
+						WalletAddress: "0xe286cbae1bbbb47d157516d334e70859a1bee4",
+						Balance:       "15",
+					},
+				}
+				return command
+			},
+			ExpectedError: ErrInvalidAddressInPrefundedAccounts,
+		},
+		{
+			TestName: "Invalid consensus engine.",
+			InvalidCommandFromValid: func(command GenesisOptions) GenesisOptions {
+				command.ConsensusEngine = "fakeConsensus"
+				return command
+			},
+			ExpectedError: ErrInvalidConsensusEngine,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.TestName, func(t *testing.T) {
+			_, err := GenerateGenesis(test.InvalidCommandFromValid(baseValidCommand))
+			if err != test.ExpectedError {
+				t.Fatalf(
+					"Invalid options did not return error. Expected error: %s, received error: %s",
+					test.ExpectedError.Error(),
+					err.Error(),
+				)
+			}
+		})
+	}
+}
+
+func TestItWritesTheGeneratedFileToAWriter(t *testing.T) {
+	cmd := GenesisOptions{
+		Network:                       "test",
+		MaxNumValidators:              "5",
+		UnbondingPeriod:               "5",
+		WalletAddressGenesisValidator: "0xe2ac86cbae1bbbb47d157516d334e70859a1bee4",
+		PrefundedAccounts: []PrefundedAccount{
+			{
+				WalletAddress: "0xe2ac86cbae1bbbb47d157516d334e70859a1bee4",
+				Balance:       "15",
+			},
+		},
+	}
+
+	generatedGenesis, err := GenerateGenesis(cmd)
+
+	if err != nil {
+		t.Fatalf("Error: %s", err.Error())
+	}
+
+	fileName := "testfiles/testnet_default.json"
```

Comment:
yeah! golden files!

add a 

`var update = flag.Bool("update", false, "update .golden files")`

and then

```
if *update {
		t.Log("update golden file")
		if err := ioutil.WriteFile(gp, b.Bytes(), 0644); err != nil {
			t.Fatalf("failed to update golden file: %s", err)
		}
	}
```

so you can generate the golden file from tests when flag is set

`go test -v ./... -update`

more details
https://medium.com/soon-london/testing-with-golden-files-in-go-7fccc71c43d3